### ISSUE TITLE
Client authentication handshake failure with Rustls

### ISSUE THREAD

**Issue by tronje** [ctz/rustls#683]
Hi! I have an issue that I *think* is specific to Rustls, although I'm prepared to be told otherwise. I'm really not entirely sure what the problem is, but here's what I know:

I'm using `warp` as my server library, and warp uses tokio-rustls. I configure and run warp like so:

```rust
warp::serve(routes)
            .tls()
            .key_path("/path/to/server.pk8")
            .cert_path("/path/to/server.crt")
            .client_auth_optional_path("/path/to/client.crt")
            .run(bind)
```

Both my client.crt and server.crt are signed by a CA cert `ca.crt`. I then use `curl` to talk to the server:

```console
$ curl -vvv --cacert ca.crt -E client.p12:foobar --cert-type P12 --tlsv1.3 https://tronje-workstation.ms.jusst.de
```

(tronje-workstation.ms.jusst.de is the domain my development machine is reachable under -- this is set correctly in the server certificate).

And I get the following error:

```
* TLSv1.3 (IN), TLS alert, handshake failure (552):
* OpenSSL SSL_read: error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure, errno 0
* Failed receiving HTTP2 data
* OpenSSL SSL_write: SSL_ERROR_ZERO_RETURN, errno 0
```

And at the same time, these logs seem to be relevant from Rustls:
```
[2021-05-04 15:52:16.932 +02:00 TRACE rustls::server] Dropping CCS
[2021-05-04 15:52:16.933 +02:00 WARN  rustls::session] Sending fatal alert HandshakeFailure
```
These show up among a lot of trace logs that I *think* are not relevant. I can provide more logs or curl output if required.

The reason I think that this has something to do with Rustls, is that the whole thing works fine with `lighttpd`.

Lighttpd config (relevant parts only):

```
ssl.privkey = "/path/to/server.pk8"
ssl.pemfile = "/path/to/server.crt"
ssl.openssl.ssl-conf-cmd = ("MinProtocol" => "TLSv1.3")
ssl.openssl.ssl-conf-cmd += ("Options" => "-ServerPreference")

ssl.verifyclient.activate = "enable"
ssl.verifyclient.enforce = "enable"
ssl.verifyclient.ca-file = "path/to/ca.crt"
```

I realize that when configuring lighttpd I give the server the CA cert, which I don't do in my warp+rustls setup. The corresponding warp API simply does not allow for this: https://docs.rs/warp/0.3.1/warp/struct.TlsServer.html
If I give warp the CA cert in place of the client cert, it panics, claming it couldn't parse the certificate, so it really seems to want a client cert here.

I suppose it may thus be a warp issue, but it was essentially a coin flip between opening the issue here or with warp. I'd greatly appreciate any insights!

**Comment by tronje** [comment]
I solved it. It was my own stupidity, really. The solution is to give warp/rustls the ca.crt and not the client.crt. I thought I couldn't do that because a) the function call is called `client_auth_required_path`, and b) because I always got a panic with my CA certificate:

```
thread 'tokio-runtime-worker' panicked at 'error binding to 0.0.0.0:7878: certificate parse error', /home/tronje/prj/warp/src/server.rs:534:27
```

(the path to warp is weird because I'm using an override for debugging)

The issue is that the file doesn't exist. I had a typo in my CLI call and didn't notice because it said "parse error". So that's it. You should use a certificate that actually exists, who could've known?

So tl;dr it was my own mistake Sorry about that, I'm closing this issue now.

**Comment by djc** [comment]
No problem, happy that you solved it!
